libglusterfs/src/inode.c

```diff
@@ -1708,11 +1708,23 @@ inode_table_with_invalidator(uint32_t lru_limit, xlator_t *xl,
         new->dentry_hashsize = dentry_hashsize;
     }
 
+    /* The size of hash table MUST always be power of 2 as required by hash_gfid
+    */
     if (inode_hashsize == 0) {
-        /* The size of hash table always should be power of 2 */
-        new->inode_hashsize = 65536;
+	new->inode_hashsize = 65536;
     } else {
-        new->inode_hashsize = inode_hashsize;
+	if (inode_hashsize < 0)
+	    inode_hashsize = 1;
+
+	if (inode_hashsize & (inode_hashsize - 1)) {
+	    /* not a power of two */
+	    inode_hashsize <<= 1;
+	    do {
+		inode_hashsize &= inode_hashsize - 1;
+	    } while (inode_hashsize & (inode_hashsize - 1));
+	}
```

Comment:
Just for academic purposes:
```suggestion
        while ((diff = inode_hashsize & (-inode_hash_size)) != inode_hash_size) {
            inode_hashsize += diff;
        }
```

Reply:
Applied locally.  Thanks.  Took me a moment to wrap my head around the math there.

Frankly, I'm back-porting this from devel onto 11 and possibly 10 branches locally.  Such a change is too big for that, and such a change belongs purely on devel.  Same applies to major reworks such as getting rid of the prev pointer, and I suggest that's handled in a separate PR entirely.

Reply:
That's fine